### PR TITLE
cephadm: add ability to specify a timeout

### PR DESCRIPTION
**Pull request by mgfritch**
Adds ability to specify a timeout during invocation of the Popen wrappers: 
- `call(..)`
- `call_throws(..)`
- `CephContainer.run(..)`
- `bootstrap.cli(..)`
- etc.

And a timeout/retry loop during `bootstrap` while waiting for services to be available:
- `mon`
- `mgr`
- `dashboard`

Also adds a global `--timeout` arg to override the default (wait forever) behavior in cases that might have been overlooked.

Signed-off-by: Michael Fritch <mfritch@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
New behavior when the `mon` fails to start during `bootstrap` ... 

```
+ ../src/ceph-daemon/ceph-daemon --image ceph/daemon-base:latest-master-devel bootstrap --mon-id a --mgr-id x --fsid 00000000-0000-0000-0000-0000deadbeef --mon-ip 127.0.0.1 --config c --output-keyring k --output-config c --allow-overwrite
INFO:root:Cluster fsid: 00000000-0000-0000-0000-0000deadbeef
INFO:ceph-daemon:Verifying we can ping mon IP 127.0.0.1...
INFO:ceph-daemon:Pulling latest ceph/daemon-base:latest-master-devel container...
INFO:ceph-daemon:Extracting ceph user uid/gid from container image...
INFO:ceph-daemon:Creating initial keys...
INFO:ceph-daemon:Creating initial monmap...
INFO:ceph-daemon:Creating mon...
INFO:ceph-daemon:Waiting for mon to start...
INFO:ceph-daemon:/usr/bin/ceph:timeout after 30 seconds
INFO:ceph-daemon:Non-zero exit code -9 from /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=bad-hostname -v /var/lib/ceph/00000000-0000-0000-0000-0000deadbeef/mon.a:/var/lib/ceph/mon/ceph-a:z -v /tmp/ceph-tmpOJeYP6:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpyREeyQ:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph ceph/daemon-base:latest-master-devel status
INFO:ceph-daemon:mon is still not available yet, waiting (1/5)...
INFO:ceph-daemon:/usr/bin/ceph:timeout after 30 seconds
INFO:ceph-daemon:Non-zero exit code -9 from /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=bad-hostname -v /var/lib/ceph/00000000-0000-0000-0000-0000deadbeef/mon.a:/var/lib/ceph/mon/ceph-a:z -v /tmp/ceph-tmpOJeYP6:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpyREeyQ:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph ceph/daemon-base:latest-master-devel status
INFO:ceph-daemon:mon is still not available yet, waiting (2/5)...
INFO:ceph-daemon:/usr/bin/ceph:timeout after 30 seconds
INFO:ceph-daemon:Non-zero exit code -9 from /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=bad-hostname -v /var/lib/ceph/00000000-0000-0000-0000-0000deadbeef/mon.a:/var/lib/ceph/mon/ceph-a:z -v /tmp/ceph-tmpOJeYP6:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpyREeyQ:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph ceph/daemon-base:latest-master-devel status
INFO:ceph-daemon:mon is still not available yet, waiting (3/5)...
INFO:ceph-daemon:/usr/bin/ceph:timeout after 30 seconds
INFO:ceph-daemon:Non-zero exit code -9 from /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=bad-hostname -v /var/lib/ceph/00000000-0000-0000-0000-0000deadbeef/mon.a:/var/lib/ceph/mon/ceph-a:z -v /tmp/ceph-tmpOJeYP6:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpyREeyQ:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph ceph/daemon-base:latest-master-devel status
INFO:ceph-daemon:mon is still not available yet, waiting (4/5)...
INFO:ceph-daemon:/usr/bin/ceph:timeout after 30 seconds
INFO:ceph-daemon:Non-zero exit code -9 from /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=bad-hostname -v /var/lib/ceph/00000000-0000-0000-0000-0000deadbeef/mon.a:/var/lib/ceph/mon/ceph-a:z -v /tmp/ceph-tmpOJeYP6:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpyREeyQ:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph ceph/daemon-base:latest-master-devel status
INFO:ceph-daemon:mon is still not available yet, waiting (5/5)...
INFO:ceph-daemon:/usr/bin/ceph:timeout after 30 seconds
INFO:ceph-daemon:Non-zero exit code -9 from /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=ceph/daemon-base:latest-master-devel -e NODE_NAME=bad-hostname -v /var/lib/ceph/00000000-0000-0000-0000-0000deadbeef/mon.a:/var/lib/ceph/mon/ceph-a:z -v /tmp/ceph-tmpOJeYP6:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpyREeyQ:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph ceph/daemon-base:latest-master-devel status
ERROR: mon is not available after 5 retries
```

File: src/ceph-daemon/ceph-daemon [hunk 1]
```diff
@@ -1322,7 +1344,11 @@ def command_run():
     c = get_container(args.fsid, daemon_type, daemon_id)
     command = c.run_cmd()
     logger.debug("Running command: %s" % ' '.join(command))
-    return subprocess.call(command)
+    try:
+        return subprocess.call(command, timeout=args.timeout) # py3
+    except TypeError:
+        # TODO: py2 has no timeout arg
+        return subprocess.call(command) # py2
```
[the remark this answers is not in the document]
Comment (pcuzner) [reply src/ceph-daemon/ceph-daemon]: if this is for octopus, is py2 support required? One thing I've done in the past is put the call in a thread that uses shell=True, and then use join with a timeout and kill the pid of the process if we hit the timeout

[the remark this answers is not in the document]
Comment (liewegas) [reply src/ceph-daemon/ceph-daemon]: ceph-daemon will run on the host, not in the container, so it needs to support py2 and py3.

File: src/ceph-daemon/ceph-daemon [hunk 2]
```diff
@@ -1149,10 +1161,13 @@ def command_bootstrap():
                 tmp_config.name: '/etc/ceph/ceph.conf:z',
             },
         )
-        out, err, ret = call(c.run_cmd(), c.entrypoint)
+        out, err, ret = call(c.run_cmd(), desc=c.entrypoint, timeout=30)
         if ret == 0:
             break
-        logger.info('mon is still not available yet, waiting...')
+        elif retry_num > retry_max:
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: Can we encapsulate this into a function or context manager? I'm seeing this loop 3 times in this PR

```python
def retry(what, retry_max, func):
        logger.info('Waiting for {what}...')
        retry_num = 1
        while True:
            if func()
                break
            elif retry_num > retry_max:
                raise Error('{what} is not available after %s tries' % retry_max)
            logger.info('{what} not yet available, waiting (%s/%s)...' % (retry_num, retry_max))
            retry_num += 1
            time.sleep(1)
```
?

**Comment by sebastian-philipp** [comment]
needs rebase

**Comment by mgfritch** [comment]
rebased

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -9,6 +9,7 @@ LOG_DIR_MODE=0o770
 DATA_DIR_MODE=0o700
 CONTAINER_PREFERENCE = ['podman', 'docker']  # prefer podman to docker
 CUSTOM_PS1=r'[ceph: \u@\h \W]\$ '
+DEFAULT_TIMEOUT=None # in seconds
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: Should we use a timeout by default or not? 

I'm a bit afraid of situations where the user has to ssh to the host to kill cephadm manually

File: src/cephadm/cephadm [hunk 4]
```diff
@@ -137,10 +138,12 @@ def call(command, desc=None, verbose=False, **kwargs):
 
     :param verbose_on_failure: On a non-zero exit status, it will forcefully set
                                logging ON for the terminal
+    :param timeout: timeout in seconds
     """
     if not desc:
         desc = command[0]
     verbose_on_failure = kwargs.pop('verbose_on_failure', True)
+    timeout = kwargs.pop('timeout', args.timeout)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: (I'd be +1 for making `timeout` a real parameter and then change this to

```suggestion
    timeout = timeout or args.timeout
```

File: src/cephadm/cephadm [hunk 5]
```diff
@@ -1323,7 +1333,7 @@ def command_bootstrap():
     tmp_config = write_tmp(config, uid, gid)
 
     # a CLI helper to reduce our typing
-    def cli(cmd, extra_mounts={}):
+    def cli(cmd, extra_mounts={}, **kwargs):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: :-1: for adding `**kwargs`` here. If we need additional parameters, we should them list explictly.

File: src/cephadm/cephadm [hunk 6]
```diff
@@ -1337,9 +1347,11 @@ def command_bootstrap():
             entrypoint='/usr/bin/ceph',
             args=cmd,
             volume_mounts=mounts,
-        ).run()
+        ).run(**kwargs)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]:
```suggestion
        ).run(timeout=timeout)
```

File: src/cephadm/cephadm [hunk 7]
```diff
@@ -1566,7 +1588,11 @@ def command_run():
     c = get_container(args.fsid, daemon_type, daemon_id)
     command = c.run_cmd()
     logger.debug("Running command: %s" % ' '.join(command))
-    return subprocess.call(command)
+    try:
+        return subprocess.call(command, timeout=args.timeout) # py3
+    except TypeError:
+        # TODO: py2 has no timeout arg
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: There are four copies of this code now. We already have a wrapper for `subprocess.call`. can we simply use this here? Alternatively we need another wrapper.

File: src/cephadm/cephadm (hunk 4 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: To be consistent perhaps `verbose_on_failure` should also be a real parameter too (and only use `kwargs` for `Popen`)?

File: src/ceph-daemon/ceph-daemon (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: https://github.com/ceph/ceph/commit/b24b2bfa5bcc7428b9392a6dfc2fabb1f334dc52#diff-8b586ec9c3ad3e8421a8858888f7ddf0R266-R276 seems to work well, thanks @pcuzner !

File: src/ceph-daemon/ceph-daemon (hunk 2 above)
Comment (mgfritch) [reply src/ceph-daemon/ceph-daemon]: Thanks! Refactored this into an `is_available` function here: https://github.com/ceph/ceph/pull/32049/commits/00028bd40ffde9ccba6d607f629dca6ef94b9f98#diff-8b586ec9c3ad3e8421a8858888f7ddf0R251-R279

File: src/cephadm/cephadm (hunk 3 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: The appropriate timeout likely varies per service. Perhaps another PR should be created to pass a default `--timeout` from `mgr/ssh` ?

File: src/cephadm/cephadm (hunk 4 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: `timeout` and `verbose_on_failure` are now real parameters

File: src/cephadm/cephadm (hunk 7 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: consolidated into a `call_timeout` function: https://github.com/ceph/ceph/commit/b24b2bfa5bcc7428b9392a6dfc2fabb1f334dc52#diff-8b586ec9c3ad3e8421a8858888f7ddf0R255-R291

**Comment by mgfritch** [comment]
retest this please

**Comment by mgfritch** [comment]
retest this please

File: src/cephadm/cephadm [hunk 8]
```diff
@@ -71,6 +73,9 @@ container_path = None
 class Error(Exception):
     pass
 
+class TimeoutExpired(Error):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: hm- we're about to create two Timeout exceptions in cephadm.

File: src/cephadm/cephadm [hunk 9]
```diff
@@ -225,14 +244,83 @@ def call(command, desc=None, verbose=False, **kwargs):
...
+    if sys.version_info >= (3, 3):
+        ret = call_timeout_py3(command, timeout)
+    else:
+        # py2 subprocess has no timeout arg
+        ret = call_timeout_py2(command, timeout)
+    return ret
+
 ##################################
 
+def is_available(what, func, args=(), kwargs=None, retry_max=5):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]:
```suggestion
def is_available(what, func, retry_max=5):
```
I'd favor to use lambdas in stead of passing args and kwargs to func. But maybe it's just me being to nit picky?

File: src/cephadm/cephadm [hunk 10]
```diff
@@ -1325,40 +1412,39 @@ def command_bootstrap():
...
+        args=[
+            'status'],
+        volume_mounts={
+            mon_dir: '/var/lib/ceph/mon/ceph-%s:z' % (mon_id),
+            tmp_admin_keyring.name: '/etc/ceph/ceph.client.admin.keyring:z',
+            tmp_config.name: '/etc/ceph/ceph.conf:z',
+        },
+    )
+    def is_mon_available(*args, **kwargs):
+        out, err, ret = call(c.run_cmd(), desc=c.entrypoint, *args, **kwargs)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]:
```suggestion
        out, err, ret = call(c.run_cmd(), desc=c.entrypoint, timeout=30)
```
Why add args and kwargs at all? But maybe I'm to nit-picky?

File: src/cephadm/cephadm (hunk 9 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: updated to use a wrapped func which seems like a nice middle ground?

**Comment by sebastian-philipp** [comment]
@mgfritch what's the state of this PR?

**Comment by mgfritch** [comment]
> @mgfritch what's the state of this PR?

rebased

**Comment by sebastian-philipp** [comment]
ok!

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-01-15_11:01:21-rados-wip-swagner-testing-distro-basic-smithi/


```
2020-01-15T13:42:20.040 INFO:tasks.workunit.client.0.smithi155.stderr:+ sudo /tmp/tmp.2XcgXyvzC6/cephadm --image ceph/daemon-base:latest-master-devel bootstrap --mon-id a --mgr-id x --mon-ip 127.0.0.1 --fsid 00000000-0000-0000-0000-0000deadbeef --config /tmp/tmp.MitILM33zp/tmp.XpX4VM5AFv --output-config /tmp/tmp.MitILM33zp/tmp.Xdy32VaTQg --output-keyring /tmp/tmp.MitILM33zp/tmp.dk1CTwA2XQ --allow-overwrite
2020-01-15T13:42:20.101 INFO:tasks.workunit.client.0.smithi155.stderr:INFO:root:Cluster fsid: 00000000-0000-0000-0000-0000deadbeef
2020-01-15T13:42:20.101 INFO:tasks.workunit.client.0.smithi155.stderr:INFO:cephadm:Verifying IP 127.0.0.1 port 3300 is not in use...
2020-01-15T13:42:20.102 INFO:tasks.workunit.client.0.smithi155.stderr:ERROR: Unable to bind to IP 127.0.0.1 port 3300:
2020-01-15T13:42:20.106 INFO:tasks.workunit.client.0.smithi155.stderr:+ rm -rf /tmp/tmp.MitILM33zp
2020-01-15T13:42:20.113 INFO:tasks.workunit.client.0.smithi155.stderr:+ rm -rf /tmp/tmp.2XcgXyvzC6
2020-01-15T13:42:20.115 DEBUG:teuthology.orchestra.run:got remote process result: 1
2020-01-15T13:42:20.115 INFO:tasks.workunit:Stopping ['cephadm/test_cephadm.sh'] on client.0...
```
seems unrelated, but we should look into making port 3300 dynamically choosen.

Rest seems fine to me. Is it g2g?

**Pull request merged**